src/helpers.c

```diff
@@ -149,130 +149,116 @@ void concat_str_array(char **output, char *const argv[])
-			char *cmdline = NULL;
+	pid_t lhspid = system_argv_fd_nowait(lnewstdinfd, pipefd[1], lnewstderrfd, &pipefd[0], lhscmd);
+	pid_t rhspid = system_argv_fd_nowait(pipefd[0], rnewstdoutfd, rnewstderrfd, &pipefd[1], rhscmd);
 
-			concat_str_array(&cmdline, argv);
-			LOG(NULL, "Failed to run command:", "%s", cmdline);
-			free(cmdline);
-		}
+	if (close(pipefd[1]) == -1) {
+		LOG(NULL, "Could not close write end of pipe file descriptor", "%d", pipefd[1]);
+		assert(0);
+	}
+	if (close(pipefd[0]) == -1) {
+		LOG(NULL, "Could not close read end of pipe file descriptor", "%d", pipefd[0]);
+		assert(0);
 	}
 
-	return status;
+	int lhsresult = wait_process_terminate(lhspid);
+	int rhsresult = wait_process_terminate(rhspid);
+	exit(rhsresult != EXIT_SUCCESS ? rhsresult : lhsresult);
 }
 
-int system_argv_fd(char *const argv[], int newstdin, int newstdout, int newstderr)
+int system_argv(char *const argv[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		if (newstdin >= 0) {
-			if (dup2(newstdin, STDIN_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdin", "");
-				assert(0);
-			}
-			close(newstdin);
-		}
-		if (newstdout >= 0) {
-			if (dup2(newstdout, STDOUT_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdout", "");
-				assert(0);
-			}
-			close(newstdout);
-		}
-		if (newstderr >= 0) {
-			if (dup2(newstderr, STDERR_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stderr", "");
-				assert(0);
-			}
-			close(newstderr);
-		}
+	return system_argv_fd(-1, -1, -1, argv);
+}
 
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
+int system_argv_fd(int newstdinfd, int newstdoutfd, int newstderrfd, char *const cmd[])
+{
+	pid_t cmdpid = system_argv_fd_nowait(newstdinfd, newstdoutfd, newstderrfd, NULL, cmd);
+	return wait_process_terminate(cmdpid);
+}
+
+pid_t system_argv_fd_nowait(int newstdinfd, int newstdoutfd, int newstderrfd, int *closefd, char *const cmd[])
+{
+	pid_t cmdpid = fork();
+	if (cmdpid == -1) {
+		LOG(NULL, "Failed to fork to execute command", "%s", cmd[0]);
 		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
+	} else if (cmdpid == 0) {
+		exec_cmd_fd(newstdinfd, newstdoutfd, newstderrfd, closefd, cmd);
+	}
+	return cmdpid;
+}
+
+void exec_cmd_fd(int newstdinfd, int newstdoutfd, int newstderrfd, int *closefd, char *const cmd[])
+{
+	move_fd(newstdinfd, STDIN_FILENO);
+	move_fd(newstdoutfd, STDOUT_FILENO);
+	move_fd(newstderrfd, STDERR_FILENO);
+	if (closefd != NULL && close(*closefd) == -1) {
+		LOG(NULL, "Could not close file descriptor", "%d", *closefd);
 		assert(0);
-	} else {
-		pid_t ws = waitpid(pid, &child_exit_status, 0);
+	}
+	execvp(*cmd, cmd);
+	LOG(NULL, "Command", "%s failed", cmd[0]);
+	assert(0);
+}
 
-		if (ws == -1) {
-			LOG(NULL, "Failed to wait for child process", "");
+void move_fd(int oldfd, int newfd)
+{
+	if (oldfd >= 0 && newfd >= 0 && oldfd != newfd) {
```

Comment:
Again this works, but I would invert the logic to "if (oldfd < 0 || newfd < 0 || oldfd == newfd ) { return ; }

Reply:
This will let us decrease indentation as well ;)  I'll also fix this.